formal-spec/Leios/Short.lagda.md

```diff
@@ -149,11 +162,23 @@ stage s = s / L
 beginningOfStage : ℕ → Type
 beginningOfStage s = stage s * L ≡ s
 
+endOfStage : ℕ → Type
+endOfStage s = suc (stage s) ≡ stage (suc s)
+
+endOfStage? : ∀ (s : ℕ) → endOfStage s ⁇
+endOfStage? s .dec = suc (stage s) ≟ stage (suc s)
+
 allDone : LeiosState → Type
-allDone record { slot = slot ; Upkeep = Upkeep } =
-      (beginningOfStage slot × Upkeep ≡ᵉ fromList (IB-Role ∷ EB-Role ∷ VT-Role ∷ Base ∷ []))
-  ⊎ (¬ beginningOfStage slot × Upkeep ≡ᵉ fromList (IB-Role ∷ VT-Role ∷ Base ∷ []))
-  ⊎ (slot ≡ zero × Upkeep ≡ᵉ fromList (Base ∷ []))
+allDone record { slot = s ; Upkeep = u ; Upkeep-Stage = v } =
```

Comment:
I feel like this has grown complicated enough to warrant a datatype?

Reply:
Yes, but like this we get the Dec type class instance for free